connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampConverter.java

```diff
@@ -326,38 +331,47 @@ protected R newRecord(R record, Schema updatedSchema, Object updatedValue) {
 
     private R applyWithSchema(R record) {
-        final Schema schema = operatingSchema(record);
+        final Schema originalSchema = operatingSchema(record);
+        final Object rawValue = operatingValue(record);
+        // Entire value is timestamp
         if (config.field.isEmpty()) {
-            Object value = operatingValue(record);
             // New schema is determined by the requested target timestamp type
-            Schema updatedSchema = TRANSLATORS.get(config.type).typeSchema();
-            return newRecord(record, updatedSchema, convertTimestamp(value, timestampTypeFromSchema(schema)));
-        } else {
-            final Struct value = requireStruct(operatingValue(record), PURPOSE);
-            Schema updatedSchema = schemaUpdateCache.get(value.schema());
-            if (updatedSchema == null) {
-                SchemaBuilder builder = SchemaUtil.copySchemaBasics(schema, SchemaBuilder.struct());
-                for (Field field : schema.fields()) {
-                    if (field.name().equals(config.field)) {
-                        builder.field(field.name(), TRANSLATORS.get(config.type).typeSchema());
-                    } else {
-                        builder.field(field.name(), field.schema());
-                    }
-                }
-                if (schema.isOptional())
-                    builder.optional();
-                if (schema.defaultValue() != null) {
-                    Struct updatedDefaultValue = applyValueWithSchema((Struct) schema.defaultValue(), builder);
-                    builder.defaultValue(updatedDefaultValue);
-                }
+            Schema updatedSchema = TRANSLATORS.get(config.type).typeSchema(originalSchema.isOptional());
+            return newRecord(record, updatedSchema, convertTimestamp(rawValue, timestampTypeFromSchema(originalSchema)));
+        }
+        // Value is Struct, only its single field should be converted
+        if (rawValue == null) {
+            Schema updatedSchema = updateSchema(originalSchema);
+            schemaUpdateCache.put(originalSchema, updatedSchema);
+            return newRecord(record, updatedSchema, null);
+        }
 
-                updatedSchema = builder.build();
-                schemaUpdateCache.put(schema, updatedSchema);
-            }
+        final Struct struct = requireStruct(rawValue, PURPOSE);
+        Schema updatedSchema = schemaUpdateCache.get(struct.schema());
+        if (updatedSchema == null) {
+            updatedSchema = updateSchema(originalSchema);
+            schemaUpdateCache.put(originalSchema, updatedSchema);
+        }
+        Struct updatedValue = applyValueWithSchema(struct, updatedSchema);
+        return newRecord(record, updatedSchema, updatedValue);
+    }
 
-            Struct updatedValue = applyValueWithSchema(value, updatedSchema);
-            return newRecord(record, updatedSchema, updatedValue);
+    private Schema updateSchema(Schema originalSchema) {
```

Comment:
Again, if we're only going to call this from one location (see previous comment), maybe it's better to not pull this logic out to minimize changes.